res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js

```diff
@@ -0,0 +1,1336 @@
+  var group = field.id.split(".")[0];
+  engine.setValue(group, "loop_out", field.value);
+}
+
+TraktorS2MK2.samplerActivateButton = function(field) {
+  if (field.value > 0) {
+    TraktorS2Mk2.controller.samplers_active[field.group] = !TraktorS2Mk2.controller.samplers_active[field.group];
+  }
+}
+
+TraktorS2MK2.connectEffectButtonLEDs = function(effectUnitGroup) {
+  TraktorS2MK2.effectButtonLEDconnections[effectUnitGroup].forEach(function(connection) {
+    connection.disconnect();
+  });
+  
+  var focusedEffect = engine.getValue(effectUnitGroup, "focused_effect");
+  var makeButtonLEDcallback = function(effectNumber) {
+    return function(value, group, control) {
+      TraktorS2MK2.controller.setOutput(effectUnitGroup, "!effectbutton" + effectNumber,
+        value === 1 ? 0x7f : 0, !TraktorS2MK2.freeze_lights);
+    };
+  };
+  
+  // FIXME: Why do the LEDs flicker?
+  TraktorS2MK2.freeze_lights = true;
+  for (var i = 0; i < 3; i++) {
+    var effectGroup;
+    var key;
+    if (focusedEffect === 0) {
+      effectGroup = effectUnitGroup.slice(0, -1) + "_Effect" + (i+1) + "]";
+      key = "enabled";
+    } else {
+      effectGroup = effectUnitGroup.slice(0, -1) + "_Effect" + focusedEffect + "]";
+      key = "button_parameter" + (i+1);
+    }
+    TraktorS2MK2.effectButtonLEDconnections[effectUnitGroup][i] = engine.makeConnection(
+      effectGroup, key, makeButtonLEDcallback(i+1));
+    TraktorS2MK2.effectButtonLEDconnections[effectUnitGroup][i].trigger();
+  }
+  TraktorS2MK2.freeze_lights = false;
+  TraktorS2MK2.effectButtonLEDconnections[effectUnitGroup][2].trigger();
+}
+
+TraktorS2MK2.onShowParametersChange = function(value, group, control) {
+  if (value === 0) {
+    if (engine.getValue(group, "show_focus") > 0) {
+      engine.setValue(group, "show_focus", 0);
+      TraktorS2MK2.previouslyFocusedEffect[group] = engine.getValue(group, "focused_effect");
+      engine.setValue(group, "focused_effect", 0);
+    }
+  } else {
+    engine.setValue(group, "show_focus", 1);
+    if (TraktorS2MK2.previouslyFocusedEffect[group] !== null) {
+      engine.setValue(group, "focused_effect", TraktorS2MK2.previouslyFocusedEffect[group]);
+    }
+  }
+  TraktorS2MK2.connectEffectButtonLEDs(group);
+}
+
+TraktorS2MK2.onFocusedEffectChange = function(value, group, control) {
+  TraktorS2MK2.controller.setOutput(group, "!effect_focus_button", value > 0 ? 0x7f : 0, !TraktorS2MK2.freeze_lights);
+  if (value === 0) {
+    for (var i = 1; i < 3; i++) {
+      // The previously focused effect is not available here, so iterate over all effects' parameter knobs.
+      for (var j = 1; j < 3; j++) {
+        engine.softTakeoverIgnoreNextValue(group.slice(0, -1) + "_Effect" + i + "]", "parameter" + j);
+      }
+    }
+  } else {
+    for (var i = 1; i < 3; i++) {
+      engine.softTakeoverIgnoreNextValue(group.slice(0, -1) + "_Effect" + i + "]", "meta");
+    }
+  }
+}
+
+TraktorS2MK2.effectFocusButton = function(field) {
```

Comment:
can you add some comments for all these functions? I'm not sure what they are doing.

Reply:
https://github.com/mixxxdj/mixxx/wiki/standard-effects-mapping

Reply:
maybe just link to that page?  
// This deck implements the standard effects mapping as described at [link]

Reply:
done